rfc/0001-oauth-access-tokens.rst

```diff
@@ -0,0 +1,345 @@
+==============
+OAuth 2.0 Authorization in the Hail Service
+==============
+
+.. author:: Daniel Goldstein
+.. date-accepted:: Leave blank. This will be filled in when the proposal is accepted.
+.. ticket-url:: Leave blank. This will eventually be filled with the
+                ticket URL which will track the progress of the
+                implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first Hail version which
+                 implements the described feature.
+.. header:: This proposal is `discussed at this pull request <https://github.com/hail-is/hail-rfc/pull/0>`_.
+            **After creating the pull request, edit this file again, update the
+            number in the link, and delete this bold sentence.**
+.. sectnum::
+.. contents::
+.. role:: python(code)
+
+Motivation
+==========
+
+This proposal focuses on the way by which users of Hail Batch and Hail Query-on-Batch
+(from now on referred to as the Hail Service) authorize programmatic access to the hail API.
+
+The Hail Service authenticates users using the OAuth2 protocol, relying on either
+GCP IAM or Azure AD as the identity providers. However, while the Hail Service
+relies on these identity providers for authentication, it currently does *not* use them
+to authorize access to Hail APIs. The Hail ``auth`` service acts as an Authorization
```

Comment:
Good question, I'm a little shaky on this but my understanding / conceptualization was: GCP/Azure act as the authentication server because they resolve a user's identity. The `authorization` server issues the credentials that are "authorized" to perform a set of actions, which is what `auth` does. It issues tokens that are authorized to perform normal user or developer operations on the Hail API.

Reply:
I think I disagree with this, but my disagreement depends on what a Hail token is:

> The Authorization Server issues the credentials that are "authorized" to perform a set of operations

In particular, I think "authorization" is a function from identity & resource to permitted/prohibited. Authorizing an action requires the ability to verify an identity, but not the ability to mint identities. A "credential", taken as plain English, sounds to me like proof of identity, which suggests to me that this server is doing authentication.

I think the key confusing thing is that Hail `auth` issues tokens. I am arguing that those tokens are indistinguishable from a proof of the user's identity. If you have that token, you can do anything the user can do. Moreover, those tokens are meant to be secret, so only the user should have that token.

I think both Google and Hail are acting as authenticators. A user hands Hail their Google proof of identity and gets a Hail proof of identity. That user can then try to take an action and Hail will authorize (or not!) that action.

I'd argue that an access token isn't a proof of identity because of how it's meant to be used: you give it to others. Hail's tokens maybe blur the line because having that token is unrestricted: it gives you every ability that user has. I suppose an important difference between an access token and a proof of identity is that a proof of identity can be used to create more access tokens but that is generally not true of an access token.

Anyway, I think what you have here is good enough for the RFC. I think, without Hail-minted tokens, auth really is an authorization only server, so let's get to that.

Reply:
Interesting, I hadn't thought about it like that. I can get behind this formulation, made the change to "Authentication Server".